File: apps/mobile/test/stored_moves_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:juchess_mobile/stored_moves.dart';

const oneMovePgn = '''[Event "?"]
[Site "?"]
[Date "????.??.??"]
[Round "?"]
[White "?"]
[Black "?"]
[Result "*"]

1. e4 *''';

void main() {
  test('PGN headers do not count as chess moves', () {
    final moves = parseStoredMoves(oneMovePgn);
    expect(moves, ['e4']);
    expect(moves.length.isOdd, isTrue, reason: 'Black must move after e4');
  });

  test('turn returns to White after Black replies', () {
    final moves = parseStoredMoves(
      oneMovePgn.replaceFirst('1. e4 *', '1. e4 e5 *'),
    );
    expect(moves, ['e4', 'e5']);
    expect(moves.length.isEven, isTrue, reason: 'White must move after e5');
  });

  test('legacy move text is parsed as legal SAN only', () {
    expect(parseStoredMoves('1. e4 {first move} e5 2. Nf3!? Nc6 *'), [
      'e4',
      'e5',
      'Nf3',
      'Nc6',
    ]);
  });
}

File: apps/web/src/lib/storedMoves.test.ts
import assert from 'node:assert/strict'
import test from 'node:test'

import { parseStoredMoves } from './storedMoves.ts'

const oneMovePgn = `[Event "?"]
[Site "?"]
[Date "????.??.??"]
[Round "?"]
[White "?"]
[Black "?"]
[Result "*"]

1. e4 *`

test('PGN headers do not count as chess moves', () => {
  assert.deepEqual(parseStoredMoves(oneMovePgn), ['e4'])
  assert.equal(parseStoredMoves(oneMovePgn).length % 2, 1, 'Black must move after White plays e4')
})

test('turn returns to White after Black replies', () => {
  const pgn = oneMovePgn.replace('1. e4 *', '1. e4 e5 *')
  assert.deepEqual(parseStoredMoves(pgn), ['e4', 'e5'])
  assert.equal(parseStoredMoves(pgn).length % 2, 0, 'White must move after Black replies e5')
})

test('legacy move text is parsed as legal SAN only', () => {
  assert.deepEqual(parseStoredMoves('1. e4 {first move} e5 2. Nf3!? Nc6 *'), [
    'e4',
    'e5',
    'Nf3',
    'Nc6',
  ])
})
